cpp/src/arrow/filesystem/gcsfs.cc

```diff
@@ -172,6 +174,79 @@ class GcsOutputStream : public arrow::io::OutputStream {
   int64_t tell_ = 0;
 };
 
+using InputStreamFactory = std::function<Result<std::shared_ptr<io::InputStream>>(
+    const std::string&, const std::string&, gcs::Generation, gcs::ReadFromOffset)>;
+
+class GcsRandomAccessFile : public arrow::io::RandomAccessFile {
+ public:
+  GcsRandomAccessFile(InputStreamFactory factory, gcs::ObjectMetadata metadata,
+                      std::shared_ptr<io::InputStream> stream)
+      : factory_(std::move(factory)),
+        metadata_(std::move(metadata)),
+        stream_(std::move(stream)) {}
+  ~GcsRandomAccessFile() override = default;
+
+  //@{
+  // @name FileInterface
+  Status Close() override { return stream_->Close(); }
+  Status Abort() override { return stream_->Abort(); }
+  Result<int64_t> Tell() const override { return stream_->Tell(); }
+  bool closed() const override { return stream_->closed(); }
+  //@}
+
+  //@{
+  // @name Readable
+  Result<int64_t> Read(int64_t nbytes, void* out) override {
+    return stream_->Read(nbytes, out);
+  }
+  Result<std::shared_ptr<Buffer>> Read(int64_t nbytes) override {
+    return stream_->Read(nbytes);
+  }
+  const arrow::io::IOContext& io_context() const override {
+    return stream_->io_context();
+  }
+  //@}
+
+  //@{
+  // @name InputStream
+  Result<std::shared_ptr<const KeyValueMetadata>> ReadMetadata() override {
+    return internal::FromObjectMetadata(metadata_);
+  }
+  //@}
+
+  //@{
+  // @name RandomAccessFile
+  Result<int64_t> GetSize() override { return metadata_.size(); }
+  Result<int64_t> ReadAt(int64_t position, int64_t nbytes, void* out) override {
+    std::shared_ptr<io::InputStream> stream;
+    ARROW_ASSIGN_OR_RAISE(stream, factory_(metadata_.bucket(), metadata_.name(),
+                                           gcs::Generation(metadata_.generation()),
+                                           gcs::ReadFromOffset(position)));
```

Comment:
Interesting. It seems a bit convoluted, but I guess the GCS C++ API is forcing us to do this.